### PR TITLE
fix: deploy vsi only to the vsi-zone-* subnets for null value

### PR DESCRIPTION
**Pull request by Aashiq-J**
### Description

the VSI is deployed to all the subnets in the VPC.
Update the logic to deploy the vsi only is vsi-zone-* subnets.

Issue: https://github.ibm.com/GoldenEye/issues/issues/6951

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Comment by ocofaigh** [comment]
@Aashiq-J I expect if we had an upgrade test for the "Existing VPC" VSI extension DA this change would actually fail the upgrade test because it will attempt to destroy VSIs from subnets that did not start with `<prefix>-vsi-zone-*`. Can we add an upgrade test and ensure that is the case? We also need to double check any documentation we may have published stating the DA deploy VSIs to all subnets, and maybe rename it to say all VSI subnets.

**Comment by Aashiq-J** [comment]
@ocofaigh ,
Added upgrade test and it is failing as expected.

It is trying to recreate the vsi in the vsi-subnets and delete the ones in the other subnets.
Screenshot 2023-12-20 at 11 39 34 AM
Screenshot 2023-12-20 at 11 39 58 AM

File: tests/pr_test.go [hunk 1]
```diff
@@ -598,3 +598,116 @@ func TestRunVsiExtention(t *testing.T) {
 		logger.Log(t, "END: Destroy (existing resources)")
 	}
 }
+
+func TestRunUpgradeVsiExtention(t *testing.T) {
```
Comment (ocofaigh) [inline tests/pr_test.go:602]: Is alot of the code here duplicated from the `TestRunVsiExtention` test? Can we maybe remove duplication by adding the common code to a function and having each test call that function?

File: patterns/vsi-extension/README.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Add a VSI to a landing zone VPC
 
-This architecture creates virtual server instances (VSI) in some or all of the subnets of one VPC of an existing landing zone deployable architecture. To create VSIs in multiple VPCs, deploy the extension once for each VPC.
+This architecture creates virtual server instances (VSI) in some or all of the VSI subnets of one VPC of an existing landing zone deployable architecture. To create VSIs in multiple VPCs, deploy the extension once for each VPC.
```
Comment (ocofaigh) [inline patterns/vsi-extension/README.md:3]: Same change needs to be made in ibm_catalog.json [here](TestRunVsiExtention)

**Comment by ocofaigh** [comment]
@Aashiq-J I wasn't expecting any new creates to be seen in the upgrade test? I thought current code was creating VSI in every subnet, and then the new code would delete the VSIs in the non VSI subnets. Why is code trying to create new VSIs?

**Comment by Aashiq-J** [comment]
@ocofaigh ,
When updating the logic to deploy only on VSI subnet, the map structure changes which terraform sees it as a new different map entries and recreates the cluster.

**Comment by Aashiq-J** [comment]
Screenshot 2024-09-23 at 12 26 39 PM

upgrade test is expected to fail because it is deleting VSIs not part of the vsi-subnets.

Also updated the override example and roks-quickstart example because these upgrade tests were also failing because it was trying to use a private endpoint for reset-api-key script.

File: patterns/vsi-extension/variables.tf [hunk 3]
```diff
@@ -100,11 +100,13 @@ variable "block_storage_volumes" {
...
-      capacity       = optional(number)
-      iops           = optional(number)
-      encryption_key = optional(string)
+      name              = string
+      profile           = string
+      capacity          = optional(number)
+      iops              = optional(number)
+      encryption_key    = optional(string)
+      resource_group_id = optional(string)
+      snapshot_id       = optional(string) # set if you would like to base volume on a snapshot
```
Comment (toddgiguere) [inline patterns/vsi-extension/variables.tf:109]: I don't think we should expose this feature yet in landing zone (the snapshot recover) as it has not been fully tested. Also when we do adopt this feature, we should also have same feature on the boot volume, and perhaps consistency group option.

For now I would just remove snapshot_id option, and when we do add it we will probably focus on that feature for that PR and release.

**Review by toddgiguere — CHANGES_REQUESTED** [review]

**Comment by Aashiq-J** [comment]
Manual upgrade test for VSI version update.
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-management-server"].ibm_is_instance.vsi["nov20-vsi-management-vsi-zone-1-0"] will be updated in-place
  ~ resource "ibm_is_instance" "vsi" {
        id                               = "02h7_188ec50c-8fdb-4f13-98f4-435becaef05c"
        name                             = "nov20-vsi-management-server-1887-001"
        tags                             = []
        # (35 unchanged attributes hidden)

      ~ primary_network_attachment {
            id            = "02h7-00c9d788-6215-4032-b0b9-6fa42e202444"
          ~ name          = "nov20-vsi-management-vsi-zone-1-eth0" -> "nov20-vsi-management-server-1887-001-vni"
            # (3 unchanged attributes hidden)

            # (1 unchanged block hidden)
        }

        # (4 unchanged blocks hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-management-server"].ibm_is_instance.vsi["nov20-vsi-management-vsi-zone-2-0"] will be updated in-place
  ~ resource "ibm_is_instance" "vsi" {
        id                               = "02i7_cfd93c44-869f-44f3-998e-bfe65272e2e3"
        name                             = "nov20-vsi-management-server-3601-001"
        tags                             = []
        # (35 unchanged attributes hidden)

      ~ primary_network_attachment {
            id            = "02i7-bbf1d3d9-209b-4645-ae3c-1caf76d829dc"
          ~ name          = "nov20-vsi-management-vsi-zone-2-eth0" -> "nov20-vsi-management-server-3601-001-vni"
            # (3 unchanged attributes hidden)

            # (1 unchanged block hidden)
        }

        # (4 unchanged blocks hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-management-server"].ibm_is_instance.vsi["nov20-vsi-management-vsi-zone-3-0"] will be updated in-place
  ~ resource "ibm_is_instance" "vsi" {
        id                               = "02j7_32fc29dc-0f22-4bda-8d8c-2d9fb451b0a7"
        name                             = "nov20-vsi-management-server-dfa6-001"
        tags                             = []
        # (35 unchanged attributes hidden)

      ~ primary_network_attachment {
            id            = "02j7-067bba04-bd7b-4e26-937d-d08dab35ca92"
          ~ name          = "nov20-vsi-management-vsi-zone-3-eth0" -> "nov20-vsi-management-server-dfa6-001-vni"
            # (3 unchanged attributes hidden)

            # (1 unchanged block hidden)
        }

        # (4 unchanged blocks hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-management-server"].ibm_is_virtual_network_interface.primary_vni["nov20-vsi-management-vsi-zone-1-0"] will be updated in-place
  ~ resource "ibm_is_virtual_network_interface" "primary_vni" {
        id                            = "02h7-78b180f0-90c5-4321-9b3b-c87159754266"
      ~ name                          = "nov20-vsi-management-vsi-zone-1-0-vni" -> "nov20-vsi-management-server-1887-001-vni"
        tags                          = []
        # (17 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-management-server"].ibm_is_virtual_network_interface.primary_vni["nov20-vsi-management-vsi-zone-2-0"] will be updated in-place
  ~ resource "ibm_is_virtual_network_interface" "primary_vni" {
        id                            = "02i7-8a06eeab-5bbf-47bd-8e44-cbd58a542708"
      ~ name                          = "nov20-vsi-management-vsi-zone-2-0-vni" -> "nov20-vsi-management-server-3601-001-vni"
        tags                          = []
        # (17 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-management-server"].ibm_is_virtual_network_interface.primary_vni["nov20-vsi-management-vsi-zone-3-0"] will be updated in-place
  ~ resource "ibm_is_virtual_network_interface" "primary_vni" {
        id                            = "02j7-53735e44-5c99-4bdf-96c9-c66851251bec"
      ~ name                          = "nov20-vsi-management-vsi-zone-3-0-vni" -> "nov20-vsi-management-server-dfa6-001-vni"
        tags                          = []
        # (17 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-workload-server"].ibm_is_instance.vsi["nov20-vsi-workload-vsi-zone-1-0"] will be updated in-place
  ~ resource "ibm_is_instance" "vsi" {
        id                               = "02h7_ee43701f-05eb-4d3f-a60c-8d1bf1e5b0e5"
        name                             = "nov20-vsi-workload-server-c013-001"
        tags                             = []
        # (35 unchanged attributes hidden)

      ~ primary_network_attachment {
            id            = "02h7-c5545b3a-675b-459d-8dfc-acea8bf10f98"
          ~ name          = "nov20-vsi-workload-vsi-zone-1-eth0" -> "nov20-vsi-workload-server-c013-001-vni"
            # (3 unchanged attributes hidden)

            # (1 unchanged block hidden)
        }

        # (4 unchanged blocks hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-workload-server"].ibm_is_instance.vsi["nov20-vsi-workload-vsi-zone-2-0"] will be updated in-place
  ~ resource "ibm_is_instance" "vsi" {
        id                               = "02i7_fb49d56c-a41f-4a75-9e8c-7340c8901447"
        name                             = "nov20-vsi-workload-server-7696-001"
        tags                             = []
        # (35 unchanged attributes hidden)

      ~ primary_network_attachment {
            id            = "02i7-7dce14c4-77fe-47a7-869a-cdea6b1d1d38"
          ~ name          = "nov20-vsi-workload-vsi-zone-2-eth0" -> "nov20-vsi-workload-server-7696-001-vni"
            # (3 unchanged attributes hidden)

            # (1 unchanged block hidden)
        }

        # (4 unchanged blocks hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-workload-server"].ibm_is_instance.vsi["nov20-vsi-workload-vsi-zone-3-0"] will be updated in-place
  ~ resource "ibm_is_instance" "vsi" {
        id                               = "02j7_86b0cb1b-dd75-4fea-affa-d2aff5727a60"
        name                             = "nov20-vsi-workload-server-bef9-001"
        tags                             = []
        # (35 unchanged attributes hidden)

      ~ primary_network_attachment {
            id            = "02j7-4d6256cf-2035-499d-b157-9536ed304cf6"
          ~ name          = "nov20-vsi-workload-vsi-zone-3-eth0" -> "nov20-vsi-workload-server-bef9-001-vni"
            # (3 unchanged attributes hidden)

            # (1 unchanged block hidden)
        }

        # (4 unchanged blocks hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-workload-server"].ibm_is_virtual_network_interface.primary_vni["nov20-vsi-workload-vsi-zone-1-0"] will be updated in-place
  ~ resource "ibm_is_virtual_network_interface" "primary_vni" {
        id                            = "02h7-9f06d58f-b054-430b-a6e1-f172c20c467c"
      ~ name                          = "nov20-vsi-workload-vsi-zone-1-0-vni" -> "nov20-vsi-workload-server-c013-001-vni"
        tags                          = []
        # (17 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-workload-server"].ibm_is_virtual_network_interface.primary_vni["nov20-vsi-workload-vsi-zone-2-0"] will be updated in-place
  ~ resource "ibm_is_virtual_network_interface" "primary_vni" {
        id                            = "02i7-063259cf-22b8-455c-9c8a-1b72eebb5ff7"
      ~ name                          = "nov20-vsi-workload-vsi-zone-2-0-vni" -> "nov20-vsi-workload-server-7696-001-vni"
        tags                          = []
        # (17 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

  # module.vsi_landing_zone.module.landing_zone.module.vsi["nov20-vsi-workload-server"].ibm_is_virtual_network_interface.primary_vni["nov20-vsi-workload-vsi-zone-3-0"] will be updated in-place
  ~ resource "ibm_is_virtual_network_interface" "primary_vni" {
        id                            = "02j7-0caa5aa7-2e0d-48be-9394-92e2d922a536"
      ~ name                          = "nov20-vsi-workload-vsi-zone-3-0-vni" -> "nov20-vsi-workload-server-bef9-001-vni"
        tags                          = []
        # (17 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 12 to change, 0 to destroy.

Changes to Outputs:
  ~ transit_gateway_data    = {
        id                      = "daccda9d-ada9-4335-911d-7a683f09a807"
        name                    = "nov20-vsi-transit-gateway"
        tags                    = []
      ~ updated_at              = "2024-11-20T09:43:59.072Z" -> "2024-11-20T09:44:03.950Z"
        # (12 unchanged attributes hidden)
    }
```
Only VNI names are renamed.

File: tests/other_test.go [hunk 4]
```diff
@@ -63,3 +74,65 @@ func TestRunOverrideExample(t *testing.T) {
 	assert.Nil(t, err, "This should not have errored")
 	assert.NotNil(t, output, "Expected some output")
 }
+
+func TestRunVsiExtention(t *testing.T) {
```
Comment (ocofaigh) [inline tests/other_test.go:78]: @Aashiq-J why was this moved to other_test.go ? Won't it mean we have a test gap in PRs where this DA is not getting tested now?

**Pull request merged**